src/types/abstract_hmm.jl

```diff
@@ -46,26 +46,29 @@ It is typically a promotion between the element type of the initialization, the
 function Base.eltype(hmm::AbstractHMM, obs, control)
     init_type = eltype(initialization(hmm))
     trans_type = eltype(transition_matrix(hmm, control))
-    dist = obs_distributions(hmm, control)[1]
-    logdensity_type = typeof(logdensityof(dist, obs))
+    dists = obs_distributions(hmm, control, obs)
+    logdensity_type = typeof(logdensityof(dists[1], obs))
     return promote_type(init_type, trans_type, logdensity_type)
 end
 
 """
     initialization(hmm)
+    initialization(hmm, control)
 
-Return the vector of initial state probabilities for `hmm`.
+Return the vector of initial state probabilities for `hmm` (possibly when `control` is applied).
```

Comment:
I don't think the initialization should depend on the control. Do you have a good argument for why it should?

Reply:
It appears in the literature, even in the article that first proposed the IOHMM ([Bengio & Frasconi, 1994](https://proceedings.neurips.cc/paper_files/paper/1994/file/8065d07da4a77621450aa84fee5656d9-Paper.pdf)). But I can give you an example:


Suppose you have two loudspeakers $A$ and $B$, and two playlists $1$ and $2$, each playlist with a set of songs $\alpha$ and $\beta$, such that $\alpha \cap \beta \neq \emptyset$ and neither $\alpha \subset \beta$ nor $\beta \subset \alpha$. I have loaded both playlists in each loudspeaker, and configured them in such a way that loudspeaker $A$ has a higher chance of playing from playlist $1$, and loudspeaker $B$ has a higher chance of playing from playlist $2$. Furthermore, if the previous song was from playlist $1$, then the next song has a higher chance of being from the same playlist, but the opposite is true from playlist $2$. The challenge goes as follows: at each time $t$, I will select one of the loudspeakers and randomly play exactly one song from one of the playlists. I can only choose the loudspeaker, but I can see the which playlist is selected and the song it played. On the other hand, you will see which loudspeaker I chose and you will hear the song, but you won't know which playlist it came from. Given what you saw and heard, can you tell me (after $N$ rounds) what is the most probable sequence of playlists?


(Of course, as a helpful friend, I'm giving you an extra information: you know exactly which songs are on $\alpha$ and $\beta$, you know that the random choice of which song in each playlist is a discrete uniform distribution and all the numerical details necessary.)


It is a silly example, I know... but it highlights one thing: the information of the first loudspeaker that I chose (`u[1]`) is very important. The initial distribution without this information would probably be of maximum entropy, however, by seeing which loudspeaker was the first, you can update your information accordingly. In fact, every information of which loudspeaker is playing the song at time $t$ is useful here, even the last one `u[N]`.